### PR TITLE
AKS setup

### PR DESCRIPTION
**Pull request by robertparkinson**
### Change description
Basic setup for AKS

### Work checklist

- [ ] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [ ] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,12 @@
-FROM node:8.9.4-alpine
+FROM hmcts/cnp-java-base:openjdk-jre-8-alpine-1.4
```
Comment (timja) [inline Dockerfile]: don't change this, its only for java.
For now update it to the latest 8.x nodejs.
After the october 1st rush, an upgrade to node 10 should be done

File: src/kubernetes/config.aat.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ${SERVICE_NAME}-config
+  namespace: ${NAMESPACE}
+data:
+  FRONTEND_BASE_URL : "https://moneyclaims.aat.platform.hmcts.net"
```
Comment (timja) [inline src/kubernetes/config.aat.yaml]: These values need to come from citizen frontend config in aat not from the claim-store value,

you can retrieve them from the app with:
```
az account set -s 1c4f0704-a29e-403d-b719-b90c34ef14c9
az webapp config appsettings list  --resource-group cmc-citizen-frontend-aat --name cmc-citizen-frontend-aat  --query '[].{ Name: name, Value: value }' -o tsv
```

make sure you don't put anything that looks like a password in here it should go in the deployment.template as a variable and it'll be retrieved from vault if you add the config in the Jenkinsfile_CNP

File: src/kubernetes/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# AKS : Azure Managed Kubernetes
+
+## Setup
+
+- Install cli: `az acs kubernetes install-cli`
+- Loging: `az login`
```
Comment (timja) [inline src/kubernetes/README.md]: typo

File: src/kubernetes/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+- Set azure subscription:  `az account set --subscription DCD-CNP-DEV`
+- Add to k8s cli (kubectl): `az aks get-credentials -n cnp-aks-cluster -g cnp-aks-rg`
```
Comment (timja) [inline src/kubernetes/README.md]: how about: `Set context to non prod AKS cluster`

File: src/kubernetes/deployment.template.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          # would usually source custom values from:
+          #   - Jenkins Azure Keyvault plugin (e.g. secrets)
+          #   - Your project's Jenkinsfile
+          env:
+            - name: REFORM_TEAM
+              value: 'cmc'
+            - name: REFORM_SERVICE_NAME
+              value: 'citizen-frontend'
+            - name: NODE_ENV
+              value: 'production'
```
Comment (timja) [inline src/kubernetes/deployment.template.yaml]: missing secrets needed here,
at least:
OAUTH_CLIENT_SECRET

File: src/kubernetes/config.aat.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  IDAM_API_URL: https://preprod-idamapi.reform.hmcts.net:3511
+  IDAM_AUTHENTICATION_WEB_URL: https://idam.preprod.ccidam.reform.hmcts.net
+  IDAM_S2S_AUTH: http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
+  PAY_URL: http://payment-api-aat.service.core-compute-aat.internal
+  CLAIM_STORE_URL: http://cmc-claim-store-aat.service.core-compute-aat.internal
+  FEES_URL: http://fees-register-api-aat.service.core-compute-aat.internal
+  DRAFT_STORE_URL: http://draft-store-service-aat.service.core-compute-aat.internal
+  FEATURE_TOGGLES_API_URL: http://rpe-feature-toggle-api-aat.service.core-compute-aat.internal
+  POSTCODE_LOOKUP_API_KEY: ${POSTCODE_LOOKUP_API_KEY}
+  APPINSIGHTS_INSTRUMENTATIONKEY: ${APPINSIGHTS_INSTRUMENTATIONKEY}
```
Comment (timja) [inline src/kubernetes/config.aat.yaml]: this should move to deployment.template.yaml, its definition isn't env specific

File: src/kubernetes/config.aat.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+data:
+  IDAM_API_URL: https://preprod-idamapi.reform.hmcts.net:3511
+  IDAM_AUTHENTICATION_WEB_URL: https://idam.preprod.ccidam.reform.hmcts.net
+  IDAM_S2S_AUTH: http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
+  PAY_URL: http://payment-api-aat.service.core-compute-aat.internal
+  CLAIM_STORE_URL: http://cmc-claim-store-aat.service.core-compute-aat.internal
+  FEES_URL: http://fees-register-api-aat.service.core-compute-aat.internal
+  DRAFT_STORE_URL: http://draft-store-service-aat.service.core-compute-aat.internal
+  FEATURE_TOGGLES_API_URL: http://rpe-feature-toggle-api-aat.service.core-compute-aat.internal
+  POSTCODE_LOOKUP_API_KEY: ${POSTCODE_LOOKUP_API_KEY}
```
Comment (timja) [inline src/kubernetes/config.aat.yaml]: this should move to deployment.template.yaml, its definition isn't env specific

Comment (markry-kainos) [reply src/kubernetes/config.aat.yaml]: i moved the UV_THREADPOOL_SIZE too - to leave only aat specific vars

**Comment by markry-kainos** [comment]
hey @timja - this looks good, can you review pls? deploy stage wont pass until cert fixed though but i've run through journey and everything fine :slightly_smiling_face: thanks for help. I'm going to update ingress and fix some env vars on claim-store now - ill ping you PR for that bit later

File: Jenkinsfile_pipeline_test [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    string(name: 'LIB_VERSION', defaultValue: 'master', description: 'Branch name of pipeline library to use')
+  ])
+])
+
+library "Infrastructure@${params.LIB_VERSION}"
+
+String product = "cmc"
+String component = "citizen-frontend"
+
+List<LinkedHashMap<String, Object>> secrets = [
```
Comment (timja) [inline Jenkinsfile_pipeline_test]: this enough vars for this to run?
if you want to keep this you could move it out to a separate PR?

File: src/kubernetes/README.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+kubectl get po -n cmc-citizen-frontend-pr-XXX  
+kubectl delete po cmc-citizen-frontend-pr-XXX-????????-?????
+
+# wait to restart and new config should have taken effect
+
+```
+
+## Known Issues
+
+- Emails will include wrong domain - you can hack a fix in similar way above. Edit configmap and change FRONTEND_BASE_URL to ingress domain for your frontend.
```
Comment (timja) [inline src/kubernetes/README.md]: is this relevant? 
claim-store is the one that will send out emails

File: src/kubernetes/test/local.yaml [hunk 10]
```diff
@@ -0,0 +1,71 @@
+
+---
```
Comment (timja) [inline src/kubernetes/test/local.yaml]: you running your app with this?
do you think this should be kept around?
interested in your thoughts

File: Jenkinsfile_pipeline_test (hunk 8 above)
Comment (markry-kainos) [reply Jenkinsfile_pipeline_test]: ill update 👍

File: src/kubernetes/README.md (hunk 9 above)
Comment (markry-kainos) [reply src/kubernetes/README.md]: not really actually - ill edit to mention though because emails will have wrong domain

File: src/kubernetes/test/local.yaml (hunk 10 above)
Comment (markry-kainos) [reply src/kubernetes/test/local.yaml]: i only used initially because i had some issues just getting application to start. it turned out it was something else but i left it in case someone else might be keen to explore?

i'm keen on your thoughts :) ive added READMEs for example where I just felt it was lacking something but its just a habit I carried over - anything that could be of use then keep in repo. my last role had zero info in the repos and i did hate it ;)

File: src/kubernetes/config.aat.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+kind: ConfigMap
+metadata:
+  name: ${SERVICE_NAME}-config
+  namespace: ${NAMESPACE}
+data:
+  IDAM_API_URL: https://preprod-idamapi.reform.hmcts.net:3511
+  IDAM_AUTHENTICATION_WEB_URL: https://idam.preprod.ccidam.reform.hmcts.net
+  IDAM_S2S_AUTH: http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
+  PAY_URL: http://payment-api-aat.service.core-compute-aat.internal
+  CLAIM_STORE_URL: http://cmc-claim-store-aat.service.core-compute-aat.internal
```
[the remark this answers is not in the document]
Comment (timja) [reply src/kubernetes/config.aat.yaml]: Hi
Please don't link dependant changes together, you should be testing in isolation it will give false confidence on a build and may cause breakages.
We don't deploy together so we shouldn't test together until the change is on master.

I'm not sure if we have formal documentation on this but every team should have been made aware of this as part of onboarding to the pipeline and any sessions where this has came up.
I did some writing on this when another team asked about it.

If more formal documentation is needed I can do it too:
https://github.com/hmcts/cnp-jenkins-library/pull/270#issuecomment-411675136

File: src/kubernetes/config.aat.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (timja) [reply src/kubernetes/config.aat.yaml]: Its up to you, 
There's talk of doing this for CCD so that teams can control callback urls...
I wouldn't do it unless you need it.
i.e. don't do it for the sake of it
If you have a real need for it then spike it out,

if you're doing that sort of thing it would be probably worth looking at helm charts and getting the teams to publish one that way you don't have all the config needed to run the services everywhere

File: src/kubernetes/config.aat.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (markry-kainos) [reply src/kubernetes/config.aat.yaml]: makes sense 👍

File: src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts [hunk 13]
```diff
@@ -1,21 +0,0 @@
-import I = CodeceptJS.I
```
Comment (timja) [inline src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: This change should probably be made as a separate PR?
its unrelated to this change

File: src/kubernetes/README.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+- Install cli: `az acs kubernetes install-cli`
+- Login: `az login`
+- Set azure subscription:  `az account set --subscription DCD-CNP-DEV`
+- Add to k8s cli (kubectl): `az aks get-credentials -n cnp-aks-cluster -g cnp-aks-rg`
+
+## Post-Setup
+
+- Set context: `kubectl config use-context cnp-aks-cluster`
+- Test by showing k8s nodes: `kubectl get nodes`
+- Show deployments, pods, services, ingress for CMC: `kubectl get deploy,po,svc,ing -n cmc`
```
Comment (timja) [inline src/kubernetes/README.md]: this is out of date, the namespace is now the pr name

File: src/kubernetes/README.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+https://tools.hmcts.net/confluence/display/CNP/Using+Containers+in+the+Pipeline#UsingContainersinthePipeline-ManualDeleteScript
+
+https://kubernetes.io/docs/reference/kubectl/cheatsheet/
+
+```
+kubectl config view                         # Show Merged kubeconfig settings.
+kubectl config current-context              # Display the current-context
+```
+
+## To Configure Frontend to use Claim-Store Backend in Kubernetes
```
Comment (timja) [inline src/kubernetes/README.md]: can you remove this section please?

File: src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts (hunk 13 above)
Comment (markry-kainos) [reply src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: this was a bit of a hack... master has been updated already

Comment (timja) [reply src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: I assume this file needs to be restored then?
This branch is up to date with master

Comment (markry-kainos) [reply src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: Can we include? :) This was so I could get a green build so then we can merge and start using AKS. The PR with this change is having some issues getting to preview env.

**Comment by markry-kainos** [comment]
Doc tweaks complete @timja - not sure if you see? I should of 'resolved' after I committed!!!!

File: src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts (hunk 13 above)
Comment (robertparkinson) [reply src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: master is currently broken, removing this file is the fix. waiting on a green build on the other PR before merging. this being merged would fix most of the fail deployments that have been delaying the merge of the other PR. happy for it to be merge in so long as there's a green build on the e2e tests

Comment (timja) [reply src/integration-test/tests/citizen/paidInFull/paid_in_full_test.ts]: aah cool, only reason I commented the above was because @markry-kainos said that the fix was on master

File: src/kubernetes/README.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+## Post-Setup
+
+- Set context: `kubectl config use-context cnp-aks-cluster`
+- Test by showing k8s nodes: `kubectl get nodes`
+- Show namespaces: `kubectl get ns`
+- Show deployments, pods, services, ingress (namespace from above): `kubectl get deploy,po,svc,ing -n cmc-citizen-frontend-pr-XXX`
+
+## Tips
+
+https://tools.hmcts.net/confluence/display/CNP/Using+Containers+in+the+Pipeline#UsingContainersinthePipeline-ManualDeleteScript
```
Comment (timja) [inline src/kubernetes/README.md]: Those docs are wrong, you only need to delete the name space now. Mind updating?
